// coc-python/lib/language/textIterator.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
const vscode_languageserver_protocol_1 = require("vscode-languageserver-protocol");
class TextIterator {
    constructor(text) {
        this.text = text;
    }
    charCodeAt(index) {
        if (index >= 0 && index < this.text.length) {
            return this.text.charCodeAt(index);
        }
        return 0;
    }
    get length() {
        return this.text.length;
    }
    getText() {
        return this.text;
    }
}
exports.TextIterator = TextIterator;
class DocumentTextIterator {
    constructor(document) {
        this.document = document;
        const lines = document.getText().split(/\r?\n/);
        const lastIndex = this.document.lineCount - 1;
        const lastLine = lines[lastIndex];
        const end = vscode_languageserver_protocol_1.Position.create(lastIndex, lastLine.length);
        this.length = this.document.offsetAt(end);
    }
    charCodeAt(index) {
        const position = this.document.positionAt(index);
        return this.document
            .getText(vscode_languageserver_protocol_1.Range.create(position, { line: position.line, character: position.character + 1 }))
            .charCodeAt(position.character);
    }
    getText() {
        return this.document.getText();
    }
}
exports.DocumentTextIterator = DocumentTextIterator;
