// coc-python/lib/interpreter/locators/services/interpreterWatcherBuilder.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
'use strict';
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
const inversify_1 = require("inversify");
const coc_nvim_1 = require("coc.nvim");
const types_1 = require("../../../common/application/types");
const logger_1 = require("../../../common/logger");
const async_1 = require("../../../common/utils/async");
const types_2 = require("../../../ioc/types");
const contracts_1 = require("../../contracts");
let InterpreterWatcherBuilder = class InterpreterWatcherBuilder {
    /**
     * Creates an instance of InterpreterWatcherBuilder.
     * Inject the DI container, as we need to get a new instance of IInterpreterWatcher to build it.
     * @param {IWorkspaceService} workspaceService
     * @param {IServiceContainer} serviceContainer
     * @memberof InterpreterWatcherBuilder
     */
    constructor(workspaceService, serviceContainer) {
        this.workspaceService = workspaceService;
        this.serviceContainer = serviceContainer;
        this.watchersByResource = new Map();
    }
    async getWorkspaceVirtualEnvInterpreterWatcher(resource) {
        const key = this.getResourceKey(resource);
        if (!this.watchersByResource.has(key)) {
            const deferred = async_1.createDeferred();
            this.watchersByResource.set(key, deferred.promise);
            const watcher = this.serviceContainer.get(contracts_1.IInterpreterWatcher, contracts_1.WORKSPACE_VIRTUAL_ENV_SERVICE);
            await watcher.register(resource);
            deferred.resolve(watcher);
        }
        return this.watchersByResource.get(key);
    }
    getResourceKey(resource) {
        const workspaceFolder = resource ? this.workspaceService.getWorkspaceFolder(resource) : undefined;
        return workspaceFolder ? coc_nvim_1.Uri.parse(workspaceFolder.uri).fsPath : '';
    }
};
__decorate([
    logger_1.traceDecorators.verbose('Build the workspace interpreter watcher')
], InterpreterWatcherBuilder.prototype, "getWorkspaceVirtualEnvInterpreterWatcher", null);
InterpreterWatcherBuilder = __decorate([
    inversify_1.injectable(),
    __param(0, inversify_1.inject(types_1.IWorkspaceService)),
    __param(1, inversify_1.inject(types_2.IServiceContainer))
], InterpreterWatcherBuilder);
exports.InterpreterWatcherBuilder = InterpreterWatcherBuilder;
