// coc-python/lib/interpreter/autoSelection/types.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.IInterpreterAutoSeletionProxyService = Symbol('IInterpreterAutoSeletionProxyService');
exports.IInterpreterAutoSelectionService = Symbol('IInterpreterAutoSelectionService');
var AutoSelectionRule;
(function (AutoSelectionRule) {
    AutoSelectionRule["all"] = "all";
    AutoSelectionRule["environment"] = "environment";
    AutoSelectionRule["currentPath"] = "currentPath";
    AutoSelectionRule["workspaceVirtualEnvs"] = "workspaceEnvs";
    AutoSelectionRule["settings"] = "settings";
    AutoSelectionRule["cachedInterpreters"] = "cachedInterpreters";
    AutoSelectionRule["systemWide"] = "system";
    AutoSelectionRule["windowsRegistry"] = "windowsRegistry";
})(AutoSelectionRule = exports.AutoSelectionRule || (exports.AutoSelectionRule = {}));
exports.IInterpreterAutoSelectionRule = Symbol('IInterpreterAutoSelectionRule');
