// coc-python/lib/common/platform/fileSystem.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
'use strict';
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const crypto_1 = require("crypto");
const fileSystem = __importStar(require("fs"));
const fs_extra_1 = __importDefault(require("fs-extra"));
const glob_1 = __importDefault(require("glob"));
const inversify_1 = require("inversify");
const path_1 = __importDefault(require("path"));
const tmp_1 = __importDefault(require("tmp"));
const async_1 = require("../utils/async");
const types_1 = require("./types");
let FileSystem = class FileSystem {
    constructor(platformService) {
        this.platformService = platformService;
    }
    get directorySeparatorChar() {
        return path_1.default.sep;
    }
    objectExists(filePath, statCheck) {
        return new Promise(resolve => {
            fs_extra_1.default.stat(filePath, (error, stats) => {
                if (error) {
                    return resolve(false);
                }
                return resolve(statCheck(stats));
            });
        });
    }
    fileExists(filePath) {
        return this.objectExists(filePath, stats => stats.isFile());
    }
    fileExistsSync(filePath) {
        return fs_extra_1.default.existsSync(filePath);
    }
    /**
     * Reads the contents of the file using utf8 and returns the string contents.
     * @param {string} filePath
     * @returns {Promise<string>}
     * @memberof FileSystem
     */
    readFile(filePath) {
        return fs_extra_1.default.readFile(filePath).then(buffer => buffer.toString());
    }
    async writeFile(filePath, data, options = { encoding: 'utf8' }) {
        await fs_extra_1.default.writeFile(filePath, data, options);
    }
    directoryExists(filePath) {
        return this.objectExists(filePath, stats => stats.isDirectory());
    }
    createDirectory(directoryPath) {
        return fs_extra_1.default.mkdirp(directoryPath);
    }
    deleteDirectory(directoryPath) {
        const deferred = async_1.createDeferred();
        fs_extra_1.default.rmdir(directoryPath, err => err ? deferred.reject(err) : deferred.resolve());
        return deferred.promise;
    }
    getSubDirectories(rootDir) {
        return new Promise(resolve => {
            fs_extra_1.default.readdir(rootDir, (error, files) => {
                if (error) {
                    return resolve([]);
                }
                const subDirs = [];
                files.forEach(name => {
                    const fullPath = path_1.default.join(rootDir, name);
                    try {
                        if (fs_extra_1.default.statSync(fullPath).isDirectory()) {
                            subDirs.push(fullPath);
                        }
                        // tslint:disable-next-line:no-empty
                    }
                    catch (ex) { }
                });
                resolve(subDirs);
            });
        });
    }
    async getFiles(rootDir) {
        const files = await fs_extra_1.default.readdir(rootDir);
        return files.filter(async (f) => {
            const fullPath = path_1.default.join(rootDir, f);
            if ((await fs_extra_1.default.stat(fullPath)).isFile()) {
                return true;
            }
            return false;
        });
    }
    arePathsSame(path1, path2) {
        path1 = path_1.default.normalize(path1);
        path2 = path_1.default.normalize(path2);
        if (this.platformService.isWindows) {
            return path1.toUpperCase() === path2.toUpperCase();
        }
        else {
            return path1 === path2;
        }
    }
    appendFileSync(filename, data, optionsOrEncoding) {
        return fs_extra_1.default.appendFileSync(filename, data, optionsOrEncoding);
    }
    getRealPath(filePath) {
        return new Promise(resolve => {
            fs_extra_1.default.realpath(filePath, (err, realPath) => {
                resolve(err ? filePath : realPath);
            });
        });
    }
    copyFile(src, dest) {
        const deferred = async_1.createDeferred();
        const rs = fs_extra_1.default.createReadStream(src).on('error', err => {
            deferred.reject(err);
        });
        const ws = fs_extra_1.default.createWriteStream(dest).on('error', err => {
            deferred.reject(err);
        }).on('close', () => {
            deferred.resolve();
        });
        rs.pipe(ws);
        return deferred.promise;
    }
    deleteFile(filename) {
        const deferred = async_1.createDeferred();
        fs_extra_1.default.unlink(filename, err => err ? deferred.reject(err) : deferred.resolve());
        return deferred.promise;
    }
    getFileHash(filePath) {
        return new Promise(resolve => {
            fs_extra_1.default.lstat(filePath, (err, stats) => {
                if (err) {
                    resolve();
                }
                else {
                    const actual = crypto_1.createHash('sha512').update(`${stats.ctimeMs}-${stats.mtimeMs}`).digest('hex');
                    resolve(actual);
                }
            });
        });
    }
    search(globPattern) {
        return new Promise((resolve, reject) => {
            glob_1.default(globPattern, (ex, files) => {
                if (ex) {
                    return reject(ex);
                }
                resolve(Array.isArray(files) ? files : []);
            });
        });
    }
    createTemporaryFile(extension) {
        return new Promise((resolve, reject) => {
            tmp_1.default.file({ postfix: extension }, (err, tmpFile, _, cleanupCallback) => {
                if (err) {
                    return reject(err);
                }
                resolve({ filePath: tmpFile, dispose: cleanupCallback });
            });
        });
    }
    createWriteStream(filePath) {
        return fileSystem.createWriteStream(filePath);
    }
    chmod(filePath, mode) {
        return new Promise((resolve, reject) => {
            fileSystem.chmod(filePath, mode, (err) => {
                if (err) {
                    return reject(err);
                }
                resolve();
            });
        });
    }
};
FileSystem = __decorate([
    inversify_1.injectable(),
    __param(0, inversify_1.inject(types_1.IPlatformService))
], FileSystem);
exports.FileSystem = FileSystem;
