// coc-python/lib/languageServices/jediProxyFactory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const coc_nvim_1 = require("coc.nvim");
const jediProxy_1 = require("../providers/jediProxy");
class JediFactory {
    constructor(extensionRootPath, serviceContainer) {
        this.extensionRootPath = extensionRootPath;
        this.serviceContainer = serviceContainer;
        this.disposables = [];
        this.jediProxyHandlers = new Map();
    }
    dispose() {
        this.disposables.forEach(disposable => disposable.dispose());
        this.disposables = [];
    }
    getJediProxyHandler(resource) {
        let workspacePath = coc_nvim_1.workspace.rootPath;
        if (!this.jediProxyHandlers.has(workspacePath)) {
            const jediProxy = new jediProxy_1.JediProxy(this.extensionRootPath, workspacePath, this.serviceContainer);
            const jediProxyHandler = new jediProxy_1.JediProxyHandler(jediProxy);
            this.disposables.push(jediProxy, jediProxyHandler);
            this.jediProxyHandlers.set(workspacePath, jediProxyHandler);
        }
        // tslint:disable-next-line:no-non-null-assertion
        return this.jediProxyHandlers.get(workspacePath);
    }
}
exports.JediFactory = JediFactory;
