// coc-python/lib/languageServices/proposeLanguageServerBanner.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
'use strict';
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
const inversify_1 = require("inversify");
const coc_nvim_1 = require("coc.nvim");
const types_1 = require("../common/application/types");
const types_2 = require("../common/types");
const random_1 = require("../common/utils/random");
const function_1 = require("../common/function");
// persistent state names, exported to make use of in testing
var ProposeLSStateKeys;
(function (ProposeLSStateKeys) {
    ProposeLSStateKeys["ShowBanner"] = "ProposeLSBanner";
})(ProposeLSStateKeys = exports.ProposeLSStateKeys || (exports.ProposeLSStateKeys = {}));
var ProposeLSLabelIndex;
(function (ProposeLSLabelIndex) {
    ProposeLSLabelIndex[ProposeLSLabelIndex["Yes"] = 0] = "Yes";
    ProposeLSLabelIndex[ProposeLSLabelIndex["No"] = 1] = "No";
    ProposeLSLabelIndex[ProposeLSLabelIndex["Later"] = 2] = "Later";
})(ProposeLSLabelIndex || (ProposeLSLabelIndex = {}));
/*
This class represents a popup that propose that the user try out a new
feature of the extension, and optionally enable that new feature if they
choose to do so. It is meant to be shown only to a subset of our users,
and will show as soon as it is instructed to do so, if a random sample
function enables the popup for this user.
*/
let ProposeLanguageServerBanner = class ProposeLanguageServerBanner {
    constructor(appShell, persistentState, configuration, sampleSizePerOneHundredUsers = 10) {
        this.appShell = appShell;
        this.persistentState = persistentState;
        this.configuration = configuration;
        this.disabledInCurrentSession = false;
        this.bannerMessage = 'Try out Preview of our new Python Language Server to get richer and faster IntelliSense completions, and syntax errors as you type.';
        this.bannerLabels = ['Try it now', 'No thanks', 'Remind me Later'];
        this.sampleSizePerHundred = sampleSizePerOneHundredUsers;
        this.initialize();
    }
    initialize() {
        if (this.initialized) {
            return;
        }
        this.initialized = true;
        // Don't even bother adding handlers if banner has been turned off.
        if (!this.enabled) {
            return;
        }
        // we only want 10% of folks that use Jedi to see this survey.
        const randomSample = random_1.getRandomBetween(0, 100);
        if (randomSample >= this.sampleSizePerHundred) {
            this.disable().catch(function_1.emptyFn);
            return;
        }
    }
    get enabled() {
        return this.persistentState.createGlobalPersistentState(ProposeLSStateKeys.ShowBanner, true).value;
    }
    async showBanner() {
        if (!this.enabled) {
            return;
        }
        const show = await this.shouldShowBanner();
        if (!show) {
            return;
        }
        const response = await coc_nvim_1.workspace.showPrompt('Use MPLS instead of jedi?');
        if (response) {
            await this.enableNewLanguageServer();
            await this.disable();
        }
    }
    async shouldShowBanner() {
        return Promise.resolve(this.enabled && !this.disabledInCurrentSession);
    }
    async disable() {
        await this.persistentState.createGlobalPersistentState(ProposeLSStateKeys.ShowBanner, false).updateValue(false);
    }
    async enableNewLanguageServer() {
        await this.configuration.updateSetting('jediEnabled', false, undefined, coc_nvim_1.ConfigurationTarget.Global);
    }
};
ProposeLanguageServerBanner = __decorate([
    inversify_1.injectable(),
    __param(0, inversify_1.inject(types_1.IApplicationShell)),
    __param(1, inversify_1.inject(types_2.IPersistentStateFactory)),
    __param(2, inversify_1.inject(types_2.IConfigurationService))
], ProposeLanguageServerBanner);
exports.ProposeLanguageServerBanner = ProposeLanguageServerBanner;
