// coc-python/lib/providers/replProvider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const coc_nvim_1 = require("coc.nvim");
const types_1 = require("../common/application/types");
const constants_1 = require("../common/constants");
const types_2 = require("../terminals/types");
class ReplProvider {
    constructor(serviceContainer) {
        this.serviceContainer = serviceContainer;
        this.disposables = [];
        this.registerCommand();
    }
    dispose() {
        this.disposables.forEach(disposable => disposable.dispose());
    }
    registerCommand() {
        const commandManager = this.serviceContainer.get(types_1.ICommandManager);
        const disposable = commandManager.registerCommand(constants_1.Commands.Start_REPL, this.commandHandler, this);
        this.disposables.push(disposable);
    }
    async commandHandler() {
        const resource = this.getActiveResourceUri();
        const replProvider = this.serviceContainer.get(types_2.ICodeExecutionService, 'repl');
        await replProvider.initializeRepl(resource);
    }
    getActiveResourceUri() {
        let doc = coc_nvim_1.workspace.getDocument(coc_nvim_1.workspace.bufnr);
        if (!doc)
            return undefined;
        return coc_nvim_1.Uri.parse(doc.uri);
    }
}
exports.ReplProvider = ReplProvider;
