// coc-python/lib/common/terminal/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var TerminalActivationProviders;
(function (TerminalActivationProviders) {
    TerminalActivationProviders["bashCShellFish"] = "bashCShellFish";
    TerminalActivationProviders["commandPromptAndPowerShell"] = "commandPromptAndPowerShell";
    TerminalActivationProviders["pyenv"] = "pyenv";
    TerminalActivationProviders["conda"] = "conda";
    TerminalActivationProviders["pipenv"] = "pipenv";
})(TerminalActivationProviders = exports.TerminalActivationProviders || (exports.TerminalActivationProviders = {}));
var TerminalShellType;
(function (TerminalShellType) {
    TerminalShellType["powershell"] = "powershell";
    TerminalShellType["powershellCore"] = "powershellCore";
    TerminalShellType["commandPrompt"] = "commandPrompt";
    TerminalShellType["gitbash"] = "gitbash";
    TerminalShellType["bash"] = "bash";
    TerminalShellType["zsh"] = "zsh";
    TerminalShellType["ksh"] = "ksh";
    TerminalShellType["fish"] = "fish";
    TerminalShellType["cshell"] = "cshell";
    TerminalShellType["tcshell"] = "tshell";
    TerminalShellType["wsl"] = "wsl";
    TerminalShellType["xonsh"] = "xonsh";
    TerminalShellType["other"] = "other";
})(TerminalShellType = exports.TerminalShellType || (exports.TerminalShellType = {}));
exports.ITerminalServiceFactory = Symbol('ITerminalServiceFactory');
exports.ITerminalHelper = Symbol('ITerminalHelper');
exports.ITerminalActivator = Symbol('ITerminalActivator');
exports.ITerminalActivationCommandProvider = Symbol('ITerminalActivationCommandProvider');
exports.ITerminalActivationHandler = Symbol('ITerminalActivationHandler');
