// coc-python/lib/application/diagnostics/types.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
var DiagnosticScope;
(function (DiagnosticScope) {
    DiagnosticScope["Global"] = "Global";
    DiagnosticScope["WorkspaceFolder"] = "WorkspaceFolder";
})(DiagnosticScope = exports.DiagnosticScope || (exports.DiagnosticScope = {}));
var DiagnosticIgnoreScope;
(function (DiagnosticIgnoreScope) {
    DiagnosticIgnoreScope["always"] = "always";
    DiagnosticIgnoreScope["session"] = "session";
})(DiagnosticIgnoreScope = exports.DiagnosticIgnoreScope || (exports.DiagnosticIgnoreScope = {}));
exports.IDiagnosticsService = Symbol('IDiagnosticsService');
exports.IDiagnosticFilterService = Symbol('IDiagnosticFilterService');
exports.IDiagnosticHandlerService = Symbol('IDiagnosticHandlerService');
exports.IInvalidPythonPathInDebuggerService = Symbol('IInvalidPythonPathInDebuggerService');
exports.ISourceMapSupportService = Symbol('ISourceMapSupportService');
