// coc-python/lib/activation/types.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.IExtensionActivationManager = Symbol('IExtensionActivationManager');
exports.IExtensionActivationService = Symbol('IExtensionActivationService');
var LanguageServerActivator;
(function (LanguageServerActivator) {
    LanguageServerActivator["Jedi"] = "Jedi";
    LanguageServerActivator["DotNet"] = "DotNet";
})(LanguageServerActivator = exports.LanguageServerActivator || (exports.LanguageServerActivator = {}));
exports.ILanguageServerActivator = Symbol('ILanguageServerActivator');
exports.IHttpClient = Symbol('IHttpClient');
exports.ILanguageServerFolderService = Symbol('ILanguageServerFolderService');
exports.ILanguageServerDownloader = Symbol('ILanguageServerDownloader');
exports.ILanguageServerPackageService = Symbol('ILanguageServerPackageService');
exports.MajorLanguageServerVersion = Symbol('MajorLanguageServerVersion');
exports.IDownloadChannelRule = Symbol('IDownloadChannelRule');
exports.ILanguageServerCompatibilityService = Symbol('ILanguageServerCompatibilityService');
var LanguageClientFactory;
(function (LanguageClientFactory) {
    LanguageClientFactory["base"] = "base";
    LanguageClientFactory["simple"] = "simple";
    LanguageClientFactory["downloaded"] = "downloaded";
})(LanguageClientFactory = exports.LanguageClientFactory || (exports.LanguageClientFactory = {}));
exports.ILanguageClientFactory = Symbol('ILanguageClientFactory');
exports.ILanguageServerAnalysisOptions = Symbol('ILanguageServerAnalysisOptions');
exports.ILanguageServerManager = Symbol('ILanguageServerManager');
exports.ILanguageServerExtension = Symbol('ILanguageServerExtension');
exports.ILanguageServer = Symbol('ILanguageServer');
var PlatformName;
(function (PlatformName) {
    PlatformName["Windows32Bit"] = "win-x86";
    PlatformName["Windows64Bit"] = "win-x64";
    PlatformName["Mac64Bit"] = "osx-x64";
    PlatformName["Linux64Bit"] = "linux-x64";
})(PlatformName = exports.PlatformName || (exports.PlatformName = {}));
exports.IPlatformData = Symbol('IPlatformData');
