// coc-python/lib/workspaceSymbols/main.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const coc_nvim_1 = require("coc.nvim");
const types_1 = require("../common/application/types");
const constants_1 = require("../common/constants");
const helpers_1 = require("../common/helpers");
const types_2 = require("../common/platform/types");
const types_3 = require("../common/process/types");
const types_4 = require("../common/types");
const generator_1 = require("./generator");
const provider_1 = require("./provider");
const MAX_NUMBER_OF_ATTEMPTS_TO_INSTALL_AND_BUILD = 2;
class WorkspaceSymbols {
    constructor(serviceContainer) {
        this.serviceContainer = serviceContainer;
        this.generators = [];
        this.outputChannel = this.serviceContainer.get(types_4.IOutputChannel, constants_1.STANDARD_OUTPUT_CHANNEL);
        this.commandMgr = this.serviceContainer.get(types_1.ICommandManager);
        this.fs = this.serviceContainer.get(types_2.IFileSystem);
        this.workspace = this.serviceContainer.get(types_1.IWorkspaceService);
        this.processFactory = this.serviceContainer.get(types_3.IProcessServiceFactory);
        this.appShell = this.serviceContainer.get(types_1.IApplicationShell);
        this.configurationService = this.serviceContainer.get(types_4.IConfigurationService);
        this.disposables = [];
        this.disposables.push(this.outputChannel);
        this.registerCommands();
        this.initializeGenerators();
        coc_nvim_1.languages.registerWorkspaceSymbolProvider(['python'], new provider_1.WorkspaceSymbolProvider(this.fs, this.commandMgr, this.generators));
        // this.disposables.push(this.workspace.onDidChangeWorkspaceFolders(() => this.initializeGenerators()))
    }
    dispose() {
        this.disposables.forEach(d => d.dispose());
    }
    initializeGenerators() {
        while (this.generators.length > 0) {
            const generator = this.generators.shift();
            generator.dispose();
        }
        if (Array.isArray(this.workspace.workspaceFolders)) {
            this.workspace.workspaceFolders.forEach(wkSpc => {
                this.generators.push(new generator_1.Generator(coc_nvim_1.Uri.parse(wkSpc.uri), this.outputChannel, this.appShell, this.fs, this.processFactory, this.configurationService));
            });
        }
    }
    registerCommands() {
        this.disposables.push(this.commandMgr.registerCommand(constants_1.Commands.Build_Workspace_Symbols, async (rebuild = true, token) => {
            const promises = this.buildWorkspaceSymbols(rebuild, token);
            return Promise.all(promises);
        }));
    }
    // tslint:disable-next-line:no-any
    buildWorkspaceSymbols(rebuild = true, token) {
        if (token && token.isCancellationRequested) {
            return [];
        }
        if (this.generators.length === 0) {
            return [];
        }
        let promptPromise;
        let promptResponse;
        return this.generators.map(async (generator) => {
            if (!generator.enabled) {
                return;
            }
            const exists = await this.fs.fileExists(generator.tagFilePath);
            // If file doesn't exist, then run the ctag generator,
            // or check if required to rebuild.
            if (!rebuild && exists) {
                return;
            }
            for (let counter = 0; counter < MAX_NUMBER_OF_ATTEMPTS_TO_INSTALL_AND_BUILD; counter += 1) {
                try {
                    await generator.generateWorkspaceTags();
                    return;
                }
                catch (error) {
                    if (!helpers_1.isNotInstalledError(error)) {
                        this.outputChannel.show();
                        return;
                    }
                }
                if (!token || token.isCancellationRequested) {
                    return;
                }
                // Display prompt once for all workspaces.
                if (promptPromise) {
                    promptResponse = await promptPromise;
                    continue;
                }
                else {
                    const installer = this.serviceContainer.get(types_4.IInstaller);
                    const uri = this.workspace.hasWorkspaceFolders ? coc_nvim_1.Uri.parse(this.workspace.workspaceFolders[0].uri) : undefined;
                    promptPromise = installer.promptToInstall(types_4.Product.ctags, uri);
                    promptResponse = await promptPromise;
                }
                if (promptResponse !== types_4.InstallerResponse.Installed || (!token || token.isCancellationRequested)) {
                    return;
                }
            }
        });
    }
}
exports.WorkspaceSymbols = WorkspaceSymbols;
