// coc-python/lib/common/application/applicationEnvironment.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
'use strict';
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
const inversify_1 = require("inversify");
const coc_nvim_1 = require("coc.nvim");
let ApplicationEnvironment = class ApplicationEnvironment {
    get appName() {
        return 'coc.nvim';
    }
    get appRoot() {
        return coc_nvim_1.workspace.pluginRoot;
    }
    get language() {
        return 'en';
    }
    get sessionId() {
        return process.pid.toString();
    }
    get machineId() {
        return '';
    }
    get extensionName() {
        // tslint:disable-next-line:non-literal-require
        return this.packageJson.name;
    }
    // tslint:disable-next-line:no-any
    get packageJson() {
        // tslint:disable-next-line:non-literal-require no-require-imports
        return require('../../../package.json');
    }
};
ApplicationEnvironment = __decorate([
    inversify_1.injectable()
], ApplicationEnvironment);
exports.ApplicationEnvironment = ApplicationEnvironment;
